### PR TITLE
Make notebook support and dependencies optional

### PR DESCRIPTION
**Pull request by Saransh-cpp**
Make nb dependencies optional

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -37,11 +37,19 @@ jobs:
...
-      - name: Test removestar
+      - name: Test removestar (lite)
+        run: |
+          pytest --cov=removestar . -vv
+
+      - name: Install removestar (nb)
+        run: |
+          python -m pip install -e ".[dev,nb]"
+
+      - name: Test removestar (lite)
```
Comment (Saransh-cpp) [inline .github/workflows/ci.yml:52]:
```suggestion
      - name: Test removestar (nb)
```

**Comment by wimglenn** [comment]
Is there a reason this is still Draft?

**Comment by Saransh-cpp** [comment]
Forgot about this. IIRC this worked and I could not extract the nb logic to a separate file. I'll merge the PR this week and create a new release.

File: removestar/__main__.py [hunk 2]
```diff
@@ -15,18 +15,16 @@
 """
 import argparse
 import glob
+import importlib
```
Comment (kieran-ryan) [inline removestar/__main__.py:18]: Believe this needs to be `import importlib.util` on [later versions of Python](https://discuss.python.org/t/python3-11-importlib-no-longer-exposes-util/25641).

```console
    and importlib.util.find_spec("nbconvert") is not None
AttributeError: module 'importlib' has no attribute 'util'
```

**Review by kieran-ryan — CHANGES_REQUESTED** [review]
Thanks for circling back on this @Saransh-cpp - much appreciated!

Just the one changed requested if you could take a look.

Separately I noticed there's a warning when IPython is not installed. Typically, this will not be observed as it installs with Jupyter, however, it will be prevalent in CI contexts where Jupyter is unlocked to be installed. It would be great if you could check whether there are different APIs or ways we could suppress this warning if we are not calling any code that could potentially trigger an error due to a missing python dependency.

```
removestar/venv/share/jupyter/nbconvert/templates/python/index.py.j2:61: UserWarning: IPython is needed to transform IPython syntax to pure Python. Install ipython if you need this functionality.
```

**Comment by Saransh-cpp** [comment]
Thanks for the review, @kieran-ryan!

I could not reproduce the warning locally. I tried uninstalling `ipython` once `jupyter` was installed, but the warning did not appear on running removestar. Could you please attach a MWE or take a look at this? Thanks!

**Comment by kieran-ryan** [comment]
@Saransh-cpp, using Python `3.11.2`, if I uninstall jupyter and ipython - and verify they are uninstalled through `pip freeze` or by attempting to run them from the command line; if I pip install `removestar 1.6.dev11+ge60ca6a` and run it against a jupyter notebook (as below), I observe the error. Would you able to advise whether applying similar reproduces the issue?

```python
from behave import *

print(step)
```

```console
➜  removestar behave.ipynb                                                          
/Users/<user>/projects/removestar/venv/share/jupyter/nbconvert/templates/python/index.py.j2:61: UserWarning: IPython is needed to transform IPython syntax to pure Python. Install ipython if you need this functionality.
--- original/behave.ipynb
+++ fixed/behave.ipynb
@@ -4,6 +4,6 @@
 # In[ ]:
 
 
-from behave import *
+from behave import step
 
 print(step)
```

**Review by kieran-ryan — APPROVED** [review]

**Comment by Saransh-cpp** [comment]
HI, @kieran-ryan, sorry for the late reply. I am not getting any time to look into the warning, but the warning looks okay to me, given that it is advising the users towards the correct usage of removestar. Could you please try resolving it if you feel otherwise? I think the warning is quite informative, and we should not suppress it.

**Comment by kieran-ryan** [comment]
No worries, thanks for reply @Saransh-cpp, I will have another look when I can.

The error relates to whether a dependency is internally installed with `removestar`, rather than relating to the user's software the package is run against; so my present thinking would be best to either suppress the warning if it does not affect functionality and if the dependency is not required; or otherwise mark it as a required dependency of `removestar` so it is installed with the package and not raised. Though will take a look and see if there's a satisfactory mitigation.

**Comment by Saransh-cpp** [comment]
Sounds good, thanks for taking a look! :)

**Comment by Saransh-cpp** [comment]
Hi, @kieran-ryan! A gentle ping. Did you make any progress on this?

**Comment by Saransh-cpp** [comment]
Hi @kieran-ryan, I tried this locally again (with fresh eyes and environment) and wasn't able to reproduce -

```py
(.env) saransh@Saranshs-MacBook-Air Projects % python3 -m pip freeze       
numpy==2.0.1
pyflakes==3.2.0
-e git+https://github.com/asmeurer/removestar.git@10365a249bb7893a48578f0d7c8b3c0b0da69ac1#egg=removestar
(.env) saransh@Saranshs-MacBook-Air Projects % removestar test.ipynb 
--- original/test.ipynb
+++ fixed/test.ipynb
@@ -4,7 +4,7 @@
 # In[ ]:
 
 
-from numpy import *
+from numpy import array
 
 a = array([1, 2, 3])
```

I will go ahead and merge this. We can fix the warning if someone reports it or if you encounter it again/

**Pull request merged**